### PR TITLE
Add CCDB manager for analysis purpose

### PR DESCRIPTION
**Pull request by njacazio**
@jgrosseo @shahor02 
- This CCDB manager limits the queries to the CCDB API
- This CCDB manager works as long as the analysis objects in the CCDB have non overlapping validity timestamp ranges

### CODE REVIEW

**Comment by shahor02** [comment]
Hi @njacazio 
How do you know a priory that the objects are non-overlapping? And if you have a way to know this (say by convention for some type of objects), then why not introducing a special option for the BasicCDBManager, it would require just a few lines. 
I see a possible problem: this singleton depends on BasicCDBManager singleton (which must be configured independently) and its cache. Some other piece of code can BasicCDBManager for the same object with different timestamp or even reset the cache, this class will never know this and, trusting to its own cached validity limits, will try to retrieve a wrong on non-existent object from the cache of BasicCDBManager. Do you agree?

**Comment by njacazio** [comment]
@shahor02 I agree with you on everything.
I tried to keep the things separate but of course the feature can go in the  `BasicCCDBManager`, it would also be easier to maintain.
If you agree I can extend the in `CachedObject` struct in `BasicCCDBManager` and port the validity feature to it, what do you think?

**Comment by shahor02** [comment]
Fine for me if you set this option off by default, but before doing this, would be good if @costing confirms that there is no way for CCDB back-end to supply object's real validity (accounting for possible overlapping updates of the same object, obviously, modulo low-probability updates after the query), this would be much safer.

File: CCDB/include/CCDB/AnalysisCCDBManager.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  std::unordered_map<std::string, CachedObject> mCache; //! map for {path, CachedObject} associations
+};
+
+template <typename T>
+T* AnalysisCCDBManager::getForTimeStamp(std::string const& path, long timestamp)
+{
+  auto& cached = mCache[path];
+  if (cached.isValid(timestamp))
+    return cached;
+  BasicCCDBManager man = BasicCCDBManager::instance();
```
Comment (ktf) [inline CCDB/include/CCDB/AnalysisCCDBManager.h:57]: Please do not use instance() everywhere, but percolate the instance in the constructor. Singletons are an anti-pattern and despite wild usage we have in O2, we should try to limit them.

File: CCDB/include/CCDB/AnalysisCCDBManager.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    long endvalidity = 0;
+    bool isValid(long ts) { return ts < endvalidity && ts > startvalidity; }
+  };
+
+ public:
+  static AnalysisCCDBManager& instance()
+  {
+    static AnalysisCCDBManager inst;
+    return inst;
+  }
```
Comment (ktf) [inline CCDB/include/CCDB/AnalysisCCDBManager.h:37]: Please avoid the singleton.

File: CCDB/include/CCDB/AnalysisCCDBManager.h (hunk 1 above)
Comment (njacazio) [reply CCDB/include/CCDB/AnalysisCCDBManager.h:57]: Hi @ktf , ok message received, I will anyway remove the new class from O2 but it is good to know

File: CCDB/include/CCDB/AnalysisCCDBManager.h (hunk 2 above)
Comment (njacazio) [reply CCDB/include/CCDB/AnalysisCCDBManager.h]: Same as before, this is now solved :)

File: CCDB/include/CCDB/BasicCCDBManager.h [hunk 3]
```diff
@@ -137,12 +147,17 @@ T* BasicCCDBManager::getForTimeStamp(std::string const& path, long timestamp)
...
+  if (mCheckObjValidityEnabled && cached.isValid(timestamp))
+    return reinterpret_cast<T*>(cached.objPtr.get());
+
   T* ptr = mCCDBAccessor.retrieveFromTFileAny<T>(path, mMetaData, timestamp, &mHeaders, cached.uuid,
                                                  mCreatedNotAfter ? std::to_string(mCreatedNotAfter) : "",
                                                  mCreatedNotBefore ? std::to_string(mCreatedNotBefore) : "");
   if (ptr) { // new object was shipped, old one (if any) is not valid anymore
     cached.objPtr.reset(ptr);
     cached.uuid = mHeaders["ETag"];
+    cached.startvalidity = std::stol(mMetaData["Valid-From"]);
```
Comment (shahor02) [inline CCDB/include/CCDB/BasicCCDBManager.h:159]: The CI tests are failing with ``std::invalid_argument: stol`` which means that ``mMetaData["Valid-From"]`` contains non-number. Could you debug the test or to run it with some logging to check what is the problem.

Comment (njacazio) [reply CCDB/include/CCDB/BasicCCDBManager.h:159]: Hi, sure thing, I'm on it

Comment (njacazio) [reply CCDB/include/CCDB/BasicCCDBManager.h]: It should be fine now, I used the Metadata instead of Headers, it is now fixed

**Comment by shahor02** [comment]
Hi @njacazio My attempt to merge has failed and testing it locally I see that that your branch does not rebase cleanly to dev:
````
git checkout dev; 
git pull;
git checkout njacazio/CCDBAnalysis
git checkout -b xxxx
git rebase dev
Applying: Add analysis CCDB manager
Applying: Add getter for last obj timestamp limits
Using index info to reconstruct a base tree...
M	CCDB/include/CCDB/BasicCCDBManager.h
Falling back to patching base and 3-way merge...
Auto-merging CCDB/include/CCDB/BasicCCDBManager.h
CONFLICT (content): Merge conflict in CCDB/include/CCDB/BasicCCDBManager.h
Recorded preimage for 'CCDB/include/CCDB/BasicCCDBManager.h'
error: Failed to merge in the changes.
Patch failed at 0002 Add getter for last obj timestamp limits
Use 'git am --show-current-patch' to see the failed patch
````
Please rebase to dev, solve the conflict and force-push again.

**Pull request merged**